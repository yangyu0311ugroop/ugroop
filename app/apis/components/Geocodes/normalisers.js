import { get, first } from 'lodash';
import { DATASTORE_UTILS } from 'datastore';
export const getLatLongs = (result, { locations }) => ({
  geocodes: locations.reduce(
    (acc, { placeId }) => ({ ...acc, [placeId]: { placeId } }),
    {},
  ),
  raw: result,
});

export const getLatLong = result => {
  const { results } = result;
  const placeId = get(results, '0.place_id', 0);
  return {
    geocodes: {
      [placeId]: {
        lat: get(results, '0.geometry.location.lat', 0),
        lng: get(results, '0.geometry.location.lng', 0),
      },
    },
  };
};

export const getCountryCode = result => {
  const { results } = result;
  const placeId = get(results, '0.place_id', null);

  if (placeId) {
    const address = get(result.results[0], 'address_components', []);
    const country = address.filter(value => value.types.includes('country'));
    const countryCode = get(first(country), 'short_name', '');
    const res = {
      [placeId]: {
        lat: get(results, '0.geometry.location.lat', 0),
        lng: get(results, '0.geometry.location.lng', 0),
        countryCode,
      },
    };
    return { geocodes: DATASTORE_UTILS.upsertObject(res) };
  }
  return null;
};

export const getCurrentLocation = result => ({ currentLocation: result });
export const GEOCODE_NORMALISERS = {
  getLatLongs,
  getLatLong,
  getCurrentLocation,
  getCountryCode,
};
