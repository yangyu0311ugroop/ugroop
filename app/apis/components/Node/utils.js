import lodash from 'lodash';
import { compose } from 'redux';
import {
  FOLDER,
  TAB_GALLERY,
  TAB_OTHER,
  TAB_TIMELINE,
  TEMPLATE,
} from 'utils/modelConstants';
import upsertHelpers, { ARRAY_MODE } from 'utils/helpers/upsertStore';
import { DATASTORE_UTILS } from 'datastore';
import { FILE_STORE_SCHEMAS } from 'datastore/fileStore/schema';
import { normalize } from 'normalizr';

export const createTemplateData = data => ({
  ...data,
  id: data.cloneId,
  people: [data.createdBy],
  children: [],
  content: data.content,
  customData: {
    shortDescription: data.shortDescription || '',
    duration: data.duration || 0,
  },
});

export const createNodeData = (type, data) => {
  switch (type) {
    case TEMPLATE: {
      return createTemplateData(data);
    }

    case FOLDER: {
      return {
        ...data,
        id: data.cloneId,
      };
    }

    default:
      return data;
  }
};

export const upsertCalculatedTimes = props => nodes => {
  props.resaga.setValue({
    nodes: upsertHelpers.deepMerge(nodes, ARRAY_MODE.SET),
  });
};

export const upsertActivityData = (node, payload) => {
  const photo = lodash.get(payload, 'node.customData.photo', undefined)
    ? {
        metaInfo: lodash.get(payload, 'node.customData.metaInfo', undefined),
        content: lodash.get(payload, 'node.customData.photo', undefined),
      }
    : null;
  const photoContent = lodash.get(photo, 'content', '');
  const customData = node.customData;
  const attachmentId = lodash.get(payload, 'node.attachment', null);
  const attachment = lodash.get(payload, 'node.customData.attachment', {});
  const attachmentDesc = lodash.get(
    payload,
    'node.customData.attachment.description',
    '',
  );

  const normalizedPhoto = photo
    ? normalize(photo, FILE_STORE_SCHEMAS.templatePhoto)
    : {};

  const attachmentOperations = attachmentId
    ? [
        DATASTORE_UTILS.upsertObject({ [attachmentId]: attachment }), // upsert attachment
        DATASTORE_UTILS.updateAttribute(
          `${attachmentId}.description`,
          attachmentDesc,
        ), // manually update description attribute
      ]
    : [];

  const photoOperations = photo
    ? [
        DATASTORE_UTILS.upsertObject(normalizedPhoto.entities.photo),
        DATASTORE_UTILS.removeObjectById(node.oldPhotoId),
      ]
    : [];

  const nodePhotoOperations = photo
    ? [
        DATASTORE_UTILS.upsertArray(`${payload.nodeId}.photos`, [photoContent]), // manually upsert photo
        DATASTORE_UTILS.upsertObject({
          [node.id]: { ...node, customData: { ...customData } },
        }),
        DATASTORE_UTILS.removeItemsArray(
          `${payload.nodeId}.photos`,
          node.oldPhotoId,
        ),
      ]
    : [];

  return {
    files: compose(...photoOperations),
    attachments: compose(...attachmentOperations),
    nodes: compose(
      DATASTORE_UTILS.upsertObject({
        [node.id]: { ...node, customData: { ...customData } },
      }),
      ...nodePhotoOperations,
    ),
  };
};

const getGalleryId = (template = {}) => {
  const { children } = template;
  if (Array.isArray(children)) {
    for (let i = 0; i < children.length; i += 1) {
      const { id, type } = children[i];

      if (type === TAB_GALLERY) {
        return id;
      }
    }
  }

  return -1;
};

const getPeopleTabId = (template = {}) => {
  const { children } = template;

  if (Array.isArray(children)) {
    for (let i = 0; i < children.length; i += 1) {
      const { id, type, customData } = children[i];

      if (type === TAB_OTHER && customData.subtype === 'tabpeople') {
        return id;
      }
    }
  }

  return -1;
};

const getTimelineId = (template = {}) => {
  const { children } = template;
  if (Array.isArray(children)) {
    for (let i = 0; i < children.length; i += 1) {
      const { id, type } = children[i];

      if (type === TAB_TIMELINE) {
        return id;
      }
    }
  }

  return -1;
};

export const NODE_API_UTILS = {
  getGalleryId,
  getTimelineId,
  getPeopleTabId,
  createTemplateData,
  createNodeData,
  upsertCalculatedTimes,
  upsertActivityData,
};
