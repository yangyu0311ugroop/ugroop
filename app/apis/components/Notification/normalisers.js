import { DATASTORE_UTILS } from 'datastore';
import { PERSON_STORE_SCHEMA } from 'datastore/personDataStore/schema';
import { normalize } from 'normalizr';
import schema from 'datastore/notificationStore/schema';
import { ORGANISATION_SCHEMA } from '../../../datastore/orgStore/schema';
const addNotifications = result => {
  const normalizedData = normalize(result, schema.notifications);
  const normalizedPeople = normalize(result, {
    users: PERSON_STORE_SCHEMA.people,
  });
  const normalizedOrg = normalize(
    result.orgs,
    ORGANISATION_SCHEMA.organisationDatas,
  );
  return {
    organisations: DATASTORE_UTILS.upsertObject(
      normalizedOrg.entities.organisationData,
    ),
    ugroopNotifications: DATASTORE_UTILS.upsertObject(
      normalizedData.entities.ugroopNotification,
    ),
    people: DATASTORE_UTILS.upsertObject(normalizedData.entities.person),
    peopleById: DATASTORE_UTILS.upsertObject(normalizedPeople.entities.person),
    nodes: DATASTORE_UTILS.upsertObject(normalizedData.entities.node),
    ugroopIds: normalizedData.result.ugroopNotifications,
    photo: DATASTORE_UTILS.upsertObject(normalizedData.entities.photo),
  };
};

export const NOTIFICATION_NORMALISERS = {
  addNotifications,
};
