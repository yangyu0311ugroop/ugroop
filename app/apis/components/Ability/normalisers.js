import { NODE_SCHEMA } from 'datastore/nodeStore/schema';
import { DATASTORE_UTILS } from 'datastore';
import first from 'lodash/first';
import { normalize } from 'normalizr';
import { compose } from 'redux';
import { helpers } from 'smartComponents/Node/logics/SortBy/config';
import { TEMPLATE } from 'utils/modelConstants';

const normalisePersonalTours = ({ myTours, sharedTours }) => {
  const { entities: myToursEntities } = normalize(
    myTours.children,
    NODE_SCHEMA.node,
  );
  const { entities: sharedToursEntities } = normalize(
    sharedTours,
    NODE_SCHEMA.node,
  );
  const personalNodes = {
    ...myToursEntities.node,
    ...sharedToursEntities.node,
  };
  const personalPhotos = {
    ...myToursEntities.photo,
    ...sharedToursEntities.photo,
  };

  const virtual = Object.keys(personalNodes)
    .reduce((accu, nodeId) => {
      if (personalNodes[nodeId].type !== TEMPLATE) return accu;

      return [...accu, [personalNodes[nodeId].updatedAt, nodeId]];
    }, [])
    .sort(helpers.sortTime());

  const sortedId = virtual.map(value => Number.parseInt(value[1], 10));
  const lastUpdatedAt = first(virtual.map(value => value[0]));

  return {
    personalNodes,
    personalPhotos,
    personalTours: {
      [-1]: {
        children: sortedId,
        updatedAt: lastUpdatedAt,
      },
    },
    updatedAt: lastUpdatedAt,
  };
};

const normaliseOrganisation = ({ orgTours, orgId }) => {
  const { entities: toursEntities } = normalize(
    orgTours[orgId].children,
    NODE_SCHEMA.node,
  );

  const virtual = Object.keys(toursEntities.node)
    .reduce((acc, nodeId) => {
      if (toursEntities.node[nodeId].type !== TEMPLATE) return acc;

      return [...acc, [toursEntities.node[nodeId].updatedAt, nodeId]];
    }, [])
    .sort(helpers.sortTime());

  const sortedId = virtual.map(value => Number.parseInt(value[1], 10));
  const lastUpdatedAt = first(virtual.map(value => value[0]));

  return {
    orgNodes: toursEntities.node,
    orgTour: {
      [orgId]: {
        children: sortedId,
        updatedAt: lastUpdatedAt,
      },
    },
    updatedAt: lastUpdatedAt,
  };
};

const cleanupTemplates = nodes =>
  Object.keys(nodes).reduce((accu, id) => {
    if (nodes[id].type === TEMPLATE) {
      const node = nodes[id];

      if (Array.isArray(node.children) && node.children.length) {
        return { ...accu, [id]: node };
      }

      delete node.children;

      return { ...accu, [id]: node };
    }

    return accu;
  }, {});

const normaliseTourRole = (accu, { nodeId, role }) => {
  const roles = accu[nodeId] && accu[nodeId].tourRoles;

  // no role yet, set the array
  if (!roles || !Array.isArray(roles))
    return {
      ...accu,
      [nodeId]: {
        tourRoles: [role],
      },
    };

  // role already exist
  if (roles.indexOf(role) !== -1) {
    return accu;
  }

  return {
    ...accu,
    [nodeId]: {
      tourRoles: roles.concat(role),
    },
  };
};

const normaliseCreatedBy = nodeData => (accu, id) => {
  if (!nodeData[id]) return accu;

  const { createdBy } = nodeData[id];

  if (createdBy && accu.indexOf(createdBy) === -1) {
    return accu.concat(createdBy);
  }

  return accu;
};

const normaliseOrganisationId = nodeData => (accu, id) => {
  if (!nodeData[id] || !nodeData[id].customData) return accu;

  const {
    customData: { organisationId },
  } = nodeData[id];

  if (organisationId && accu.indexOf(organisationId) === -1) {
    return accu.concat(organisationId);
  }

  return accu;
};

const findMyTours = ({ myTours, orgTours, sharedTours, userNodes }) => {
  const orgs = Object.keys(orgTours).reduce(
    (accu, orgId) => {
      if (!orgTours[orgId].children || !orgTours[orgId].children.length)
        return {
          ...accu,
          sorts: [...accu.sorts, [0, orgId]],
        };

      const { orgNodes, orgTour, updatedAt } = normaliseOrganisation({
        orgTours,
        orgId,
      });

      return {
        nodes: { ...accu.nodes, ...orgNodes },
        orgNodes: { ...accu.orgNodes, ...orgTour },
        sorts: [...accu.sorts, [updatedAt, orgId]],
      };
    },
    { nodes: {}, orgNodes: {}, sorts: [] },
  );
  const myOrgTourIds = Object.keys(orgs.nodes);

  // remove tours from org you are member with
  const filteredShareTours = sharedTours.filter(
    ({ id = 0 }) => !myOrgTourIds.includes(id.toString()),
  );

  const {
    personalNodes,
    personalPhotos,
    personalTours,
    updatedAt: personalUpdatedAt,
  } = normalisePersonalTours({ myTours, sharedTours: filteredShareTours });

  const sortedOrganisationIds = [[personalUpdatedAt, -1], ...orgs.sorts]
    .sort(helpers.sortTime())
    .map(value => Number.parseInt(value[1], 10));

  const cleanupNodes = cleanupTemplates({
    ...orgs.nodes,
    ...personalNodes,
  });

  const tourRoles = userNodes.reduce(ABILITY_NORMALISERS.normaliseTourRole, {});

  const allNodes = compose(
    DATASTORE_UTILS.upsertObject(tourRoles),
    DATASTORE_UTILS.upsertObject(cleanupNodes),
  );

  const nodeData = allNodes({});
  const nodeIds = Object.keys(nodeData).map(value =>
    Number.parseInt(value, 10),
  );

  const userIds = Object.keys(nodeData).reduce(
    ABILITY_NORMALISERS.normaliseCreatedBy(nodeData),
    [],
  );

  const organisationIds = Object.keys(nodeData).reduce(
    ABILITY_NORMALISERS.normaliseOrganisationId(nodeData),
    [],
  );
  return {
    files: DATASTORE_UTILS.upsertObject({
      ...personalPhotos,
    }),
    node: compose(
      DATASTORE_UTILS.upsertObject(tourRoles),
      DATASTORE_UTILS.upsertObject(cleanupNodes),
    ),
    organisationTours: compose(
      DATASTORE_UTILS.upsertObject(personalTours),
      DATASTORE_UTILS.upsertObject(orgs.orgNodes),
    ),
    sortedOrganisationIds,
    require: { userIds, organisationIds, nodeIds },
  };
};

export const ABILITY_NORMALISERS = {
  normaliseCreatedBy,
  normaliseOrganisationId,
  normaliseTourRole,
  findMyTours,
  normalisePersonalTours,
  normaliseOrganisation,
  cleanupTemplates,
};
