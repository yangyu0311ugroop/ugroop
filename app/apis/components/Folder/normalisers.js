import { FOLDER_FETCH_TYPES } from 'apis/components/Folder/constants';
import { convertChecklists } from 'apis/components/Template/config';
import { normalize } from 'normalizr';
import { compose } from 'redux';
import { DATASTORE_UTILS } from 'datastore';
import { NODE_NORMALISERS, NODE_SCHEMA } from 'datastore/nodeStore/schema';
import { NODE_TRAIL_NORMALISERS } from 'datastore/nodeTrailsDataStore/schema';
import dotProp from 'dot-prop-immutable';
import union from 'lodash/union';
import difference from 'lodash/difference';
import { get } from 'lodash';
import { CHECKLIST, FOLDER } from 'utils/modelConstants';

const getChecklists = result => {
  let nodes;

  if (result.type === FOLDER) {
    nodes = dotProp.set(result, 'checklists', checklists =>
      checklists.map(checklist =>
        dotProp.set(checklist, 'checklists', convertChecklists),
      ),
    );
  } else if (result.type === CHECKLIST) {
    nodes = dotProp.set(result, 'checklists', convertChecklists);
  } else {
    nodes = result;
  }

  const { node } = NODE_NORMALISERS.normalise(nodes);
  if (result.returnNode) {
    return node;
  }
  return { nodes: DATASTORE_UTILS.upsertObject(node) };
};

const mergeFolder = (id, folder) => (store = {}) => {
  if (Object.keys(store).length > 0) {
    const oldFolderChildren = dotProp.get(
      store,
      `${id}.calculated.sortedIds`,
      [],
    );
    const newFolderChildren = dotProp.get(
      folder,
      `${id}.calculated.sortedIds`,
      [],
    );
    const mergedFolderChildren = union(oldFolderChildren, newFolderChildren);
    const mergedFolder = dotProp.set(
      folder,
      `${id}.calculated.sortedIds`,
      mergedFolderChildren,
    );
    return dotProp.merge(store, `${id}`, mergedFolder[id]);
  }

  return DATASTORE_UTILS.upsertObject(folder)(store);
};

const fetchFolderChildren = (nodeList, payload = {}) => {
  const normalizedData = NODE_NORMALISERS.folderNormalise(nodeList);
  const count = get(nodeList, 'calculated.sortedIds.length');

  const insertFolder =
    payload.fetchType === FOLDER_FETCH_TYPES.PAGINATION
      ? mergeFolder(normalizedData.id, normalizedData.folder)
      : DATASTORE_UTILS.upsertObject(normalizedData.folder);

  return {
    nodes: compose(
      DATASTORE_UTILS.upsertObject(normalizedData.nodes),
      insertFolder,
    ),
    files: DATASTORE_UTILS.upsertObject(normalizedData.files),
    people: DATASTORE_UTILS.upsertObject(normalizedData.people),
    nodeId: normalizedData.id,
    count,
  };
};

const fetchFolderParent = (nodes, payload) => {
  const restructuredObj = { id: parseInt(payload, 10), trail: nodes };
  const normalizedData = NODE_TRAIL_NORMALISERS.breadcrumbNormaliser(
    restructuredObj,
  );

  return {
    trail: normalizedData.trail,
    breadcrumb: normalizedData.breadcrumb,
    nodes: DATASTORE_UTILS.upsertObject(normalizedData.nodes),
  };
};

export const getUnexistingNodesInStore = (normalizedResult, store) => {
  const addedNodes = Object.keys(normalizedResult.entities.node);
  const currNodes = Object.keys(store);
  const differenceBetweenCurrAndAdded = difference(addedNodes, currNodes);
  const toBeAddedNodes = differenceBetweenCurrAndAdded.map(
    id => normalizedResult.entities.node[id],
  );

  return normalize(toBeAddedNodes, NODE_SCHEMA.node);
};

const batchFetchFolderTree = result => {
  const normalizedResult = normalize(result, NODE_SCHEMA.node);

  return {
    nodes: nodes => {
      const normalizedFilteredResult = getUnexistingNodesInStore(
        normalizedResult,
        nodes,
      );
      return DATASTORE_UTILS.upsertObject(
        normalizedFilteredResult.entities.node,
      )(nodes);
    },
    raw: result,
  };
};

const handleBatchClone = (result = []) => {
  const finalValue = result.map(val =>
    getChecklists({ ...val.customVal, returnNode: true }),
  );
  const cloneIds = result.map(val => val.cloneId);
  const node = finalValue.reduce((acc, curr) => ({ ...acc, ...curr }), {});
  return { nodes: DATASTORE_UTILS.upsertObject(node), result, cloneIds };
};

export default {
  getChecklists,
  fetchFolderChildren,
  fetchFolderParent,
  mergeFolder,
  batchFetchFolderTree,
  handleBatchClone,
};
