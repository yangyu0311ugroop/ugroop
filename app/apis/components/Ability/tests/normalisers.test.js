import { FOLDER, GUEST, TEMPLATE, TOUR_ORGANIZER } from 'utils/modelConstants';
import { ABILITY_NORMALISERS } from '../normalisers';

describe('Ability/config.js', () => {
  beforeEach(() => jest.clearAllMocks());

  const mockResponse = {
    myTours: {
      id: 1,
      type: FOLDER,
      children: [
        {
          id: 112,
          type: TEMPLATE,
          children: [],
          content: 'tour',
          updatedAt: '112',
        },
        {
          id: 11,
          type: FOLDER,
          content: 'folder',
          children: [
            {
              id: 113,
              type: TEMPLATE,
              children: [],
              content: 'tour 2',
              updatedAt: '113',
              photos: [{ id: 999, content: 'some photo' }],
            },
          ],
        },
      ],
    },
    orgTours: {
      2233: {
        id: 2222,
        type: FOLDER,
        children: [
          {
            id: 221,
            type: TEMPLATE,
            children: [],
            content: 'tour 3',
            updatedAt: '221',
          },
          {
            id: 22,
            type: FOLDER,
            content: 'folder',
            children: [
              {
                id: 223,
                type: TEMPLATE,
                children: [],
                content: 'tour 4',
                updatedAt: '223',
              },
            ],
          },
        ],
      },
      2234: {
        children: [],
      },
      2235: {},
    },
    sharedTours: [
      {
        id: 334,
        type: TEMPLATE,
        children: [],
        content: 'tour 5',
        updatedAt: '334',
      },
      {
        id: 335,
        type: TEMPLATE,
        children: [],
        content: 'tour 6',
        updatedAt: '335',
      },
    ],
    userNodes: [
      { nodeId: 334, role: TOUR_ORGANIZER },
      { nodeId: 335, role: GUEST },
    ],
    orgId: 2233,
  };

  describe('ABILITY_NORMALISERS()', () => {
    describe('normalisePersonalTours()', () => {
      it('should normalisePersonalTours()', () => {
        expect(
          ABILITY_NORMALISERS.normalisePersonalTours(mockResponse),
        ).toMatchSnapshot();
      });
    });

    describe('normaliseOrganisation()', () => {
      it('should normaliseOrganisation()', () => {
        expect(
          ABILITY_NORMALISERS.normaliseOrganisation(mockResponse),
        ).toMatchSnapshot();
      });
    });

    describe('normaliseTourRole()', () => {
      it('should normaliseTourRole()', () => {
        expect(
          ABILITY_NORMALISERS.normaliseTourRole(
            {},
            { nodeId: 2233, role: 'tour_viewer' },
          ),
        ).toEqual({ 2233: { tourRoles: ['tour_viewer'] } });
      });

      it('should normaliseTourRole() 2', () => {
        expect(
          ABILITY_NORMALISERS.normaliseTourRole(
            { 2233: { tourRoles: ['tour_owner'] } },
            { nodeId: 2233, role: 'tour_viewer' },
          ),
        ).toEqual({ 2233: { tourRoles: ['tour_owner', 'tour_viewer'] } });
      });

      it('should normaliseTourRole() 3', () => {
        expect(
          ABILITY_NORMALISERS.normaliseTourRole(
            { 2233: { tourRoles: ['tour_viewer'] } },
            { nodeId: 2233, role: 'tour_viewer' },
          ),
        ).toEqual({ 2233: { tourRoles: ['tour_viewer'] } });
      });
    });

    describe('normaliseCreatedBy()', () => {
      it('should normaliseCreatedBy()', () => {
        expect(
          ABILITY_NORMALISERS.normaliseCreatedBy({
            4455: { createdBy: 33 },
          })([55], 3344),
        ).toEqual([55]);
      });

      it('should normaliseCreatedBy() 1', () => {
        expect(
          ABILITY_NORMALISERS.normaliseCreatedBy({
            4455: { createdBy: 33 },
          })([], 4455),
        ).toEqual([33]);
      });

      it('should normaliseCreatedBy() 2', () => {
        expect(
          ABILITY_NORMALISERS.normaliseCreatedBy({
            4455: { createdBy: 33 },
          })([33], 4455),
        ).toEqual([33]);
      });
    });

    describe('normaliseOrganisationId()', () => {
      it('should normaliseOrganisationId()', () => {
        expect(
          ABILITY_NORMALISERS.normaliseOrganisationId({
            4455: { customData: { organisationId: 33 } },
          })([55], 3344),
        ).toEqual([55]);
      });

      it('should normaliseOrganisationId() 1', () => {
        expect(
          ABILITY_NORMALISERS.normaliseOrganisationId({
            4455: { customData: { organisationId: 33 } },
          })([], 4455),
        ).toEqual([33]);
      });

      it('should normaliseOrganisationId() 2', () => {
        expect(
          ABILITY_NORMALISERS.normaliseOrganisationId({
            4455: { customData: { organisationId: 33 } },
          })([33], 4455),
        ).toEqual([33]);
      });
    });

    describe('cleanupTemplates()', () => {
      it('should cleanupTemplates()', () => {
        expect(
          ABILITY_NORMALISERS.cleanupTemplates({
            1: { type: TEMPLATE, children: [] },
            2: { type: FOLDER, children: [1] },
            3: { type: TEMPLATE, children: [1, 2] },
          }),
        ).toMatchSnapshot();
      });
    });

    describe('findMyTours()', () => {
      it('should findMyTours()', () => {
        ABILITY_NORMALISERS.normaliseTourRole = jest.fn(() => () => ({}));
        ABILITY_NORMALISERS.normaliseCreatedBy = jest.fn(() => () => ({}));
        ABILITY_NORMALISERS.normaliseOrganisationId = jest.fn(() => () => ({}));
        const result = ABILITY_NORMALISERS.findMyTours(mockResponse);
        expect(result.sortedOrganisationIds.length).toEqual(4);
      });
    });
  });
});
