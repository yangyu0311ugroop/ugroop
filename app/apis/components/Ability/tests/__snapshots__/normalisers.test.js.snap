// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Ability/config.js ABILITY_NORMALISERS() cleanupTemplates() should cleanupTemplates() 1`] = `
Object {
  "1": Object {
    "type": "template",
  },
  "3": Object {
    "children": Array [
      1,
      2,
    ],
    "type": "template",
  },
}
`;

exports[`Ability/config.js ABILITY_NORMALISERS() normaliseOrganisation() should normaliseOrganisation() 1`] = `
Object {
  "orgNodes": Object {
    "22": Object {
      "children": Array [
        223,
      ],
      "content": "folder",
      "id": 22,
      "type": "folder",
    },
    "221": Object {
      "children": Array [],
      "content": "tour 3",
      "id": 221,
      "type": "template",
      "updatedAt": "221",
    },
    "223": Object {
      "children": Array [],
      "content": "tour 4",
      "id": 223,
      "type": "template",
      "updatedAt": "223",
    },
  },
  "orgTour": Object {
    "2233": Object {
      "children": Array [
        223,
        221,
      ],
      "updatedAt": "223",
    },
  },
  "updatedAt": "223",
}
`;

exports[`Ability/config.js ABILITY_NORMALISERS() normalisePersonalTours() should normalisePersonalTours() 1`] = `
Object {
  "personalNodes": Object {
    "11": Object {
      "children": Array [
        113,
      ],
      "content": "folder",
      "id": 11,
      "type": "folder",
    },
    "112": Object {
      "children": Array [],
      "content": "tour",
      "id": 112,
      "type": "template",
      "updatedAt": "112",
    },
    "113": Object {
      "children": Array [],
      "content": "tour 2",
      "id": 113,
      "photos": Array [
        "some photo",
      ],
      "type": "template",
      "updatedAt": "113",
    },
    "334": Object {
      "children": Array [],
      "content": "tour 5",
      "id": 334,
      "type": "template",
      "updatedAt": "334",
    },
    "335": Object {
      "children": Array [],
      "content": "tour 6",
      "id": 335,
      "type": "template",
      "updatedAt": "335",
    },
  },
  "personalPhotos": Object {
    "some photo": Object {
      "content": "some photo",
      "id": 999,
    },
  },
  "personalTours": Object {
    "-1": Object {
      "children": Array [
        335,
        334,
        113,
        112,
      ],
      "updatedAt": "335",
    },
  },
  "updatedAt": "335",
}
`;
